### DISCUSSION TITLE
New Bulit-in Service: Group Chat.

### DISCUSSION THREAD

**Discussion by sunhuachuang** [CympleTech/esse#5] [Ideas]
## What ?
Group chat service in ESSE.

## Why ?
We need not only one-way communication between people, but also group communication, but this kind of group communication, different from one-way communication, requires a third-party group owner to provide a communication platform.

## How ?
1. Group Chat Agreement.
2. Implementation of the client (ESSE) of the group chat protocol.
3. The server implementation of the group chat protocol.

Three group chat type:
- **Encrypted**: data **encrypted**, and need group member invite or manager agree.
- **Private**: data **not encrypted**, and need group member invite or manager agree.
- **Open**: data **not encrypted**, and anyone can join this group chat.

First discuss the encryption situation, all data in the group service is encrypted, and group members share a **Symmetrically encrypted
Key**, and group service keep the **Key's hash** value, when group service check if someone can join the group/get group message, he must provide a **zero knowledge proof** about the **Key**, so group service can verify it.

**Comment by tiannian** [comment]
I have two questions:
1. How to generate and distribute symmetrically encrypted key. To keep forward safety, we must rotation key in period.
2. How to use zhproof to proof key? To proof you have this key?

**Reply by sunhuachuang** [reply]
1. Encrypted key, not public share, only when you invite someone, add to the invite link (invite chat message).
2. this proof circuit is simple:
    - first: `copy/eq` about current Peer network ID (PeerID, not user did) gate.
    - second: `hash proof` gate.
